lib/puppet/provider/keystore_certificate/openssl.rb

```diff
@@ -0,0 +1,95 @@
+Puppet::Type.type(:keystore_certificate).provide(:openssl) do
+  commands :openssl => 'openssl'
+  commands :keytool => 'keytool'
+
+  def create
+    add_certificate
+  end
+
+  def destroy
+    delete_certificate
+  end
+
+  def exists?
+    keystore_content
+  end
+
+  def certificate
+    keystore_fingerprint
+  end
+
+  def certificate=(value)
+    delete_certificate unless keystore_content.nil?
+    add_certificate
+  end
+
+  def fingerprint(file)
+    return unless File.exist?(file)
+
+    openssl('x509', '-sha256', '-noout', '-fingerprint', '-in', file).strip.split('=')[1]
+  end
+
+  private
+
+  def add_certificate
+    Tempfile.open('temp_keystore') do |temp_store|
+      openssl(
+        'pkcs12',
+        '-export',
+        '-in', resource[:certificate],
+        '-inkey', resource[:private_key],
+        '-out', temp_store,
+        '-name', resource[:alias],
+        '-CAfile', resource[:ca],
+        '-password', "file:#{resource[:password_file]}"
+      )
+
+      keytool(
+        '-importkeystore',
+        '-noprompt',
+        '-srckeystore', temp_store,
+        '-srcstorepass:file', resource[:password_file],
+        '-destkeystore', resource[:keystore],
+        '-deststorepass:file', resource[:password_file],
+        '-srcalias', resource[:alias],
+        '-destalias', resource[:alias]
+      )
+    end
+  rescue Puppet::ExecutionFailure => e
+    Puppet.err("Failed to add certificate to keystore: #{e}")
+    nil
+  end
+
+  def delete_certificate
+    keytool(
+      '-delete',
+      '-noprompt',
+      '-keystore', resource[:keystore],
+      '-alias', resource[:alias],
+      '-storepass:file', resource[:password_file]
+    )
+  end
+
+  def keystore_content
+    return unless file_readable?(resource[:keystore])
+    return unless file_readable?(resource[:password_file])
+
+    keytool(
+      '-list',
+      '-keystore', resource[:keystore],
+      '-storepass:file', resource[:password_file],
+      '-alias', resource[:alias]
+    )
+  rescue Puppet::ExecutionFailure => e
+    Puppet.debug("Failed to read keystore contents: #{e}")
+    nil
+  end
+
+  def keystore_fingerprint
+    keystore_content&.scan(/^Certificate fingerprint \(SHA-256\): (.+)$/)&.first&.first
```

Comment:
After a bit of testing, I understand that the second `&.first` is necessary because otherwise idempotence is broken:

```
Notice: /Stage[main]/Certs::Candlepin/Keystore_certificate[/etc/candlepin/certs/keystore:tomcat]/certificate: certificate changed ['FD:64:E2:53:2D:2D:30:AC:8D:C7:9A:D4:9C:D8:DD:D1:3D:09:98:B5:41:9D:F0:ED:66:9F:40:C2:7B:13:7C:7E'] to 'FD:64:E2:53:2D:2D:30:AC:8D:C7:9A:D4:9C:D8:DD:D1:3D:09:98:B5:41:9D:F0:ED:66:9F:40:C2:7B:13:7C:7E'
```

I had a concern about multiple certificates in the keystore, but we are specifying the `-alias` so we should get only the one we care about

Reply:
That's the idea.